### ISSUE TITLE
[Bug] Segfault from SDL2 when a PS4 controller abruptly disconnects

### ISSUE THREAD

**Issue by lostromb** [Ryujinx/Ryujinx#4453]
### Description of the issue

I am running DS4Windows in the background to manage some DualShock 4 controllers. Normally this would expose them as XInput controllers to games, but because SDL2 also natively supports the devices, a single controller setup up in this way shows up on the Ryujinx input tab simultaneously as an Xbox360 controller and a PS4 controller. However, this appears to cause some kind of conflicting handle situation between Ryujinx, SDL, and DS4Windows which ends up leading to a segfault. This is the only capture I could get of the error as it is not logged and only shows up briefly in the console.

![image](https://user-images.githubusercontent.com/918053/220013153-a56709a6-73a1-458f-bc92-f242102b8d26.png)

The workaround for this currently is to use the controller in XInput mode. But it would be nice to not have potential fatal crashes lingering around.

### Reproduction steps

- Run DS4Windows and pair a DualShock 4 controller, using output mapping "Auto"
- Run Ryujinx and configure input to use that controller as a PS4 controller (not XInput)
- Start a game
- In DS4Windows, either configure an idle disconnect for the controller and wait for it to time out, or it's faster to just modify the controller's profile and save it, which triggers an abrupt disconnect+reconnect and also reproes this bug in Ryujinx. A regular disconnect of the controller (though PS button + options button combo) doesn't trigger the bug.

### Log file

The error is a process segfault and does not reach the logs

### OS

Windows 11

### Ryujinx version

1.1.626

**Comment by lostromb** [comment]
hmmmm.... can't repro it on my dev machine. that's fun

**Comment by Squall-Leonhart** [comment]
Disable exclusive mode in ds4 windows.

**Comment by lostromb** [comment]
> Disable exclusive mode in ds4 windows.

It's not using exclusive mode, as far as I can tell.
```
02/28/2023 10:05:13: Starting...
02/28/2023 10:05:13: Using output KB+M handler: SendInput
02/28/2023 10:05:13: Connection to ViGEmBus 1.16.112.0 established
02/28/2023 10:05:13: Searching for controllers...
02/28/2023 10:05:13: Using Shared Mode
02/28/2023 10:05:13: UDP server listening on address 127.0.0.1 port 26760
02/28/2023 10:06:55: Found Controller: AC:FD:93:CD:BC:A1 (BT) (DS4 v.1).
02/28/2023 10:06:55: Plugging in virtual X360 controller (XInput slot #?) in output slot #1
02/28/2023 10:06:55: Associated input controller #1 (DS4 v.1) to virtual X360 Controller in output slot #1
02/28/2023 10:06:55: Controller 1 is using Profile "Default". Battery level is 100%
02/28/2023 11:00:40: AC:FD:93:CD:BC:A1 disconnecting due to idle disconnect
02/28/2023 11:00:40: Disassociated virtual X360 Controller in output slot #1 from input controller #1 (DS4 v.1)
02/28/2023 11:00:40: Unplugging virtual X360 Controller from output slot #1
02/28/2023 11:00:40: Controller 1 was removed or lost connection
```

**Comment by gdkchan** [comment]
It could be a double free issue, have you checked if `SDL_GameControllerClose` is being called twice with the same pointer?

**Comment by lostromb** [comment]
> It could be a double free issue, have you checked if `SDL_GameControllerClose` is being called twice with the same pointer?

Yeah, it looks like that's the case. If I add a basic log line to track each pad's handle I get:
```
Acquired SDL gamepad 204037A9FB0
Releasing SDL gamepad 204037A9FB0
Acquired SDL gamepad 204037AF688
Releasing SDL gamepad 204037AF688
Acquired SDL gamepad 204037AA248
Releasing SDL gamepad 204037AA248
Acquired SDL gamepad 204037A9FB0
Releasing SDL gamepad 204037A9FB0
Releasing SDL gamepad 204037AA248 <<< ???
Acquired SDL gamepad 204037A9FB0
Acquired SDL gamepad 204037AA248
Releasing SDL gamepad 204037A9FB0
Releasing SDL gamepad 204037A9FB0 <<< ???
Releasing SDL gamepad 204037AA248
```

Perhaps something gets mixed up when the input settings window scans for all controllers and finds two entries that correspond to the same physical device?

Just for sanity I'm going to try adding a proper atomic `isDisposed` lock to `SDL2Gamepad.Dispose()` to see if it's a simple race condition from multiple disposal

**Comment by Squall-Leonhart** [comment]
Sounds like a variation of https://github.com/Ryujinx/Ryujinx/issues/3301
